File: 2015/R/d07.R
parse <- function(f) {
  d <- read.csv(text = gsub(" -> ", " ", 
    gsub("NOT", "0 NOT", readLines(f))), sep = " ",
    header = FALSE, col.names = c("arg1", "op", "arg2", "dest"))
  assigns <- which(d$dest == "")
  d$dest[assigns] <- d$op[assigns]
  d$op[assigns] <- "EQ"
  d$arg2[assigns] <- 0
  d$stage <- 0
  d$stage[numberish(d$arg1) & numberish(d$arg2)] <- 1
  d[, c("dest", "op", "arg1", "arg2", "stage")]
}

numberish <- function(s) suppressWarnings(!is.na(as.numeric(s)))
RSHIFT <- function(x, y) bitwShiftR(as.numeric(x), as.numeric(y))
LSHIFT <- function(x, y) bitwShiftL(as.numeric(x), as.numeric(y)) 
NOT <- function(x, y) bitwNot(as.numeric(y))
AND <- function(x, y) bitwAnd(as.numeric(x), as.numeric(y))
OR <- function(x, y) bitwOr(as.numeric(x), as.numeric(y))
EQ <- function(x, y) as.numeric(x)

solve <- function(d) {
  while(any(d$stage < 2)) {
    is <- which(d$stage == 1)
    for (i in is) {
      res <- do.call(d$op[i], list(d$arg1[i], d$arg2[i]))
      d$arg1[d$arg1 == d$dest[i]] <- res
      d$arg2[d$arg2 == d$dest[i]] <- res
      d$stage[i] <- 2
    }
    d$stage <- ifelse(d$stage == 0, numberish(d$arg1) & numberish(d$arg2),
                      d$stage)
  }
  res <- which(d$dest == "a")
  do.call(d$op[res], list(d$arg1[res], d$arg2[res]))
}

d <- parse("../inputs/input_7.txt")
p1 <- solve(d)
d$arg1[d$dest == 'b'] <- p1
c(p1, solve(d))
                  

File: 2015/python/d07.py
def read(filename):
    for line in open(filename, 'r'):
        bits = line.replace('\n', '').split(' -> ')
        lhs = bits[0].split(' ')
        if len(lhs) == 3:
            swap = lhs[0]
            lhs[0] = lhs[1]
            lhs[1] = swap
        facts[bits[1]] = lhs

def calc(sym):
    if (sym.isdigit()):
        return int(sym)
    if (sym in cache):
        return cache[sym]
    eq = facts[sym]
    if (len(eq) == 1):
        if (eq[0].isdigit()):
            res = int(eq[0])
        else:
            res = calc(eq[0])
    elif (eq[0] == 'NOT'):
        res = 65535 - calc(eq[1])
    elif (eq[0] == 'AND'):
        res = calc(eq[1]) & calc(eq[2])
    elif (eq[0] == 'OR'):
        res = calc(eq[1]) | calc(eq[2])
    elif (eq[0] == 'LSHIFT'):
        res = calc(eq[1]) << calc(eq[2])
    elif (eq[0] == 'RSHIFT'):
        res = calc(eq[1]) >> calc(eq[2])
    cache[sym] = res
    return res

def solve(lhs):
    return calc(lhs)

def tests():
    facts = {}
    cache = {}
    read('../Java/d07/test.txt')
    solve('d')
    solve('e')
    solve('f')
    solve('g')
    solve('h')
    solve('i')
    assert(cache['d'] == 72)
    assert(cache['e'] == 507)
    assert(cache['f'] == 492)
    assert(cache['g'] == 114)
    assert(cache['h'] == 65412)
    assert(cache['i'] == 65079)
    assert(cache['x'] == 123)
    assert(cache['y'] == 456)

facts = {}
cache = {}
read('../inputs/input_7.txt')
a = solve('a')
print(a)
facts["b"] = [str(a)]
cache = {}
print(solve('a'))
